pkg/kubelet/images/image_pull_manager.go

```diff
@@ -0,0 +1,663 @@
+func (n *NamedLockSet) GlobalUnlock() {
+	for _, l := range n.locks {
+		l.Unlock()
+	}
+	n.globalLock.Unlock()
+}
+
+// FileBasedImagePullManager is an implementation of the ImagePullManager. It
+// tracks images pulled by the kubelet by creating records about ongoing and
+// successful pulls.
+// It tracks the credentials used with each successful pull in order to be able
+// to distinguish tenants requesting access to an image that exists on the kubelet's
+// node.
+type FileBasedImagePullManager struct {
+	pullingDir string
+	pulledDir  string
+
+	imagePullPolicy ImagePullPolicyEnforcer
+
+	imageService kubecontainer.ImageService
+
+	intentAccessors *NamedLockSet  // image -> sync.Mutex
+	intentCounters  map[string]int // image -> number of current in-flight pulls
+
+	pulledAccessors *NamedLockSet // imageRef -> sync.Mutex
+
+	encoder runtime.Encoder
+	decoder runtime.Decoder
+}
+
+func NewFileBasedImagePullManager(ctx context.Context, kubeletDir string, imagePullPolicy ImagePullPolicyEnforcer, imageService kubecontainer.ImageService) (*FileBasedImagePullManager, error) {
+	kubeletConfigEncoder, kubeletConfigDecoder, err := createKubeletConfigSchemeEncoderDecoder()
+	if err != nil {
+		return nil, err
+	}
+
+	m := &FileBasedImagePullManager{
+		pullingDir: filepath.Join(kubeletDir, "image_manager", "pulling"),
+		pulledDir:  filepath.Join(kubeletDir, "image_manager", "pulled"),
+
+		imagePullPolicy: imagePullPolicy,
+
+		imageService: imageService,
+
+		intentAccessors: NewNamedLockSet(),
+		intentCounters:  make(map[string]int),
+
+		pulledAccessors: NewNamedLockSet(),
+
+		encoder: kubeletConfigEncoder,
+		decoder: kubeletConfigDecoder,
+	}
+
+	if err := os.MkdirAll(m.pullingDir, 0700); err != nil {
+		return nil, err
+	}
+
+	if err := os.MkdirAll(m.pulledDir, 0700); err != nil {
+		return nil, err
+	}
+
+	m.startupHousekeeping(ctx)
+
+	return m, nil
+}
+
+func (f *FileBasedImagePullManager) RecordPullIntent(image string) error {
+	f.intentAccessors.Lock(image)
+	defer f.intentAccessors.Unlock(image)
+
+	intent := imagemanagerv1alpha1.ImagePullIntent{
+		Image: image,
+	}
+
+	intentBytes := bytes.NewBuffer([]byte{})
+	if err := f.encoder.Encode(&intent, intentBytes); err != nil {
+		return err
+	}
+
+	if err := writeFile(f.pullingDir, cacheFilename(image), intentBytes.Bytes()); err != nil {
+		return err
+	}
+
+	f.intentCounters[image]++
+	return nil
+}
+
+func (f *FileBasedImagePullManager) RecordImagePulled(image, imageRef string, credentials *imagemanagerv1alpha1.ImagePullCredentials) {
+	if err := f.writePulledRecord(image, imageRef, credentials); err != nil {
+		klog.ErrorS(err, "failed to write image pulled record", "imageRef", imageRef)
+		return
+	}
+
+	f.dereferenceImagePullIntent(image)
+}
+
+// writePulledRecord writes an ImagePulledRecord into the f.pulledDir directory.
+func (f *FileBasedImagePullManager) writePulledRecord(image, imageRef string, credentials *imagemanagerv1alpha1.ImagePullCredentials) error {
```

Comment:
document what it means if credentials is nil (we call it this way in initialize and error cases)